extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+# workspace/reference extension to LSP
+
+The `workspace/reference` extension to the Language Server Protocol (LSP) enables a language server to export all of the references that code inside of the workspace root makes to code symbols that are external to the workspace root (e.g. dependencies).
+
+Use case: clients of a language server can invoke `workspace/reference` in order to find external references to e.g. imported dependencies. This information can then be stored inside of a database, which allows the caller to create a 'global mapping' of symbols in dependencies to the workspace they are used in (e.g. to see "how do other people use this symbol?").
+
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ServerCapabilities {
+  /* ... all fields from the base ServerCapabilities ... */
+
+  /**
+   * The server provides workspace reference exporting support.
+   */
+  workspaceReferenceProvider?: boolean;
+}
+```
+
+#### Workspace References Request
+
+The workspace reference request is sent from the client to the server to export project-wide external references. The response strictly includes only the locations in the workspace at which symbols _outside_ of the workspace are referenced. That is:
+
+- Excluding any `ReferenceInformation.URI` which is located within the workspace.
+- Excluding any `ReferenceInformation.Name` which is unexported or private (i.e. according to language semantics).
+- Excluding any `ReferenceInformation.Location` which is located in vendored code (e.g. `vendor/...` for Go, `node_modules/...` for JS, .tgz NPM packages, or .jar files for Java).
+
+
+_Request_
+* method: 'workspace/reference'
+* params: `WorkspaceReferenceParams` defined as follows:
+```typescript
+/**
+ * The parameters of a Workspace Reference Request.
+ */
+interface WorkspaceReferenceParams {
+	/**
+	 * A limit on the number of returned results.
+	 */
+	limit: number
+}
+```
+
+_Response_:
+* result: `ReferenceInformation[]` defined as follows:
+```typescript
+/**
+ * Represents information about a reference to programming constructs like
+ * variables, classes, interfaces etc.
+ */
+interface ReferenceInformation {
+	/**
+	 * The location in the workspace at which the `symbol` has been referenced.
+	 */
+	location: Location;
+
+	/**
+	 * The name of the symbol that is being referenced. For example:
+	 *
+	 * 	(Go) "ServeHTTP"
+	 *  (JS) "render"
+	 *
+	 */
+	name?: string;
+
+	/**
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
```

Comment:
In PHP, `uri`, `name` and `containerName` is not enough to uniquely identify a symbol.
Example:
```php
class Test
{
  const test = 123;
  public function test() {}
}
```
Both symbols share the same `uri`, `name` (test) and `containerName` (Test). You can make it unique by adding in the `kind` attribute from `SymbolInformation`, because one is a constant, and the other is a method.
In fact, I build a fully qualified name for every symbol of indexing purposes. The FQNs for the constant is `Test::test`, for the method it's `Test::test()`. It could get even more complicated in languages like Java that allow overloading. In that case you would need to include the signature of the method (parameters and return value). So if you need a unique identifier, imo it should be done as an additional property (like `fqn`). The LS can then choose how we builds these fully qualified names in a way that makes sense for the language. It might even improve the UX for Sourcegraph if you expose this to the user, because a PHP user is familiar with the syntax of separating class and method with a double colon, while a Java/JavaScript user is familiar with separating them with a dot.